### ISSUE TITLE
0.0.76 regression: filtered list comprehension incorrectly inferred as list[Never]

### ISSUE THREAD

**Issue by jelle-openai** [astral-sh/ty#4432]
### Summary

This standalone example passes with **ty 0.0.75** and fails with **0.0.76**.

```python
from typing import Any

class A: ...

def f0(v0: Any) -> list[A]:
    v1 = [A() for _, v2 in enumerate(getattr(v0, 'r8', ())) if 'q9' in v2.a0 and v2.a1]
    v1.append(A())
    return v1
```

Save as `repro.py` and run:

```sh
ty check --python-version 3.12 repro.py
```

**Expected:** the comprehension produces `A` instances, so appending an `A` and returning `list[A]` should be accepted.

**Actual on 0.0.76:** exit 1, with two diagnostics:

```text
error[invalid-argument-type]: Argument to bound method `list.append` is incorrect
Expected `Never`, found `A`

error[invalid-return-type]: Return type does not match returned value
expected `list[A]`, found `list[Never]`
```

This is not necessarily an empty or unreachable comprehension. Using `types.SimpleNamespace`, calling `f0(SimpleNamespace(r8=[SimpleNamespace(a0=["q9"], a1=True)]))` successfully returns two `A` instances at runtime.

Two controls pass with both versions:

- Explicitly annotate `v1: list[A]`.
- Remove `and v2.a1` from the filter.

Removing `enumerate` and its unused index also makes the corresponding reduction pass in both versions. I have not bisected the regression.

Tested with the official macOS ARM64 binaries, Python 3.12 typing mode:

```text
ty 0.0.75 (a553dcc1a 2026-08-26)
ty 0.0.76 (1940c8a75 2026-08-31)
```

**Comment by carljm** [comment]
This is another case of losing gradual evidence in constraint solving; astral-sh/ruff#26873 fixes it. The underlying bug has been there forever (`enumerate(x)` where `x: Any | tuple[()]` reproduces it more simply); it was exposed in this example by astral-sh/ruff#28061, which preserved `Never` more accurately in unpacking assignments instead of falling back to `Unknown`.
